### PR TITLE
Postcondition function with return value

### PR DESCRIPTION
**Pull request by feliperodri**
Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

*Description of changes:*

Define a function contract to encapsulate the check of a postcondition with the return value of a function. The cost in prod is 0, because when the `DEBUG` mode is turn off the condition is not even evaluated.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/common/common.h [hunk 1]
```diff
@@ -41,6 +41,19 @@
...
+/**
+ * Define a function contract to encapsulate the check of a postcondition with
+ * the return value of a function. The cost in prod is 0, because when the
+ * DEBUG mode is turn off the condition is not even evaluated.
+ */
+#ifdef NDEBUG
+#define AWS_RETURN_SUCCESS_WITH_POSTCONDITION(cond) (AWS_OP_SUCCESS)
+#define AWS_RETURN_SIZE_WITH_POSTCONDITION(cond, size) (size)
+#else
+#define AWS_RETURN_SUCCESS_WITH_POSTCONDITION(cond) aws_assert_postcondition(cond)
```
Comment (danielsn) [inline include/aws/common/common.h]: Cleaner way
```
/* Use the comma operator to evaluate the assert, then return the value
 * https://en.wikipedia.org/wiki/Comma_operator
 */
#define AWS_RETURN_VALUE_WITH_POSTCONDITION(val, cond) (assert(cond), (val))
#define AWS_RETURN_SUCCESS_WITH_POSTCONDITION(cond)  AWS_RETURN_VALUE_WITH_POSTCONDITION(AWS_OP_SUCCESS, (cond))
```

Comment (feliperodri) [reply include/aws/common/common.h]: great, I'm on it!

File: include/aws/common/common.h [hunk 2]
```diff
@@ -41,6 +41,20 @@
                                     abort();                                                                    \
                                }                                                                                \
 
+/**
+ * Define a function contract to encapsulate the check of a postcondition with
+ * the return value of a function. The cost in prod is 0, because when the
+ * DEBUG mode is turn off the condition is not even evaluated. This keep the
+ * assert in the same line for debuggability.
+ */
+#ifdef NDEBUG
```
Comment (danielsn) [inline include/aws/common/common.h]: You don't need the ndebug check anymore, since the assert will already handle that.

File: include/aws/common/common.h [hunk 3]
```diff
@@ -50,20 +50,24 @@
...
 #endif
 
-
+/**
+ * Define a function contract to encapsulate the check of a postcondition with
+ * the return value of a function. This keep the assert in the same line for debuggability.
+ * If NDEBUG is on, this will have 0 runtime cost.
+ */
+#define AWS_RETURN_VALUE_WITH_POSTCONDITION(cond, val) (assert(cond), (val))
+#define AWS_RETURN_SUCCESS_WITH_POSTCONDITION(cond) AWS_RETURN_VALUE_WITH_POSTCONDITION((cond), AWS_OP_SUCCESS)
```
Comment (danielsn) [inline include/aws/common/common.h:70]: What macro should we use for void functions?

Comment (feliperodri) [reply include/aws/common/common.h:70]: `AWS_POSTCONDITION` only? As we currently have

File: include/aws/common/array_list.inl [hunk 4]
```diff
@@ -80,9 +80,8 @@ bool aws_array_list_is_valid(const struct aws_array_list *AWS_RESTRICT list) {
     size_t required_size;
     bool required_size_is_valid = aws_mul_size_checked(list->length, list->item_size, &required_size) == AWS_OP_SUCCESS;
     bool current_size_is_valid = list->current_size >= required_size;
-    bool item_size_is_valid = list->item_size != 0;
```
Comment (tautschnig) [inline include/aws/common/array_list.inl:83]: Unrelated change?

**Review by tautschnig — COMMENTED** [review]
I'm worried about the (seemingly) unrelated change. Otherwise it seems ok, but not super convincing to me: I personally find the existing code easier to read.

File: .cbmc-batch/jobs/aws_array_list_capacity/aws_array_list_capacity_harness.c [hunk 5]
```diff
@@ -16,36 +16,28 @@
 #include <aws/common/array_list.h>
 #include <proof_helpers/make_common_data_structures.h>
 
-/* These values allow us to reach a higher coverage rate */
-#define MAX_ITEM_SIZE 2
-#define MAX_INITIAL_ITEM_ALLOCATION (UINT64_MAX / MAX_ITEM_SIZE) + 1
-
 /**
- * Runtime: 0m4.951s
+ * Runtime: 5s
```
Comment (markrtuttle) [inline .cbmc-batch/jobs/aws_array_list_capacity/aws_array_list_capacity_harness.c:20]: Still not sure this data belongs in the source code, seems like a maintenance headache.

File: .cbmc-batch/jobs/aws_array_list_back/cbmc-batch.yaml [hunk 6]
```diff
@@ -1,4 +1,4 @@
 jobos: ubuntu16
-cbmcflags: "--bounds-check;--pointer-check;--div-by-zero-check;--signed-overflow-check;--unsigned-overflow-check;--pointer-overflow-check;--undefined-shift-check;--float-overflow-check;--nan-check;--unwinding-assertions;--unwindset;memcpy_impl.0:3;--function;aws_array_list_back_harness"
+cbmcflags: "--bounds-check;--pointer-check;--div-by-zero-check;--signed-overflow-check;--unsigned-overflow-check;--pointer-overflow-check;--undefined-shift-check;--float-overflow-check;--nan-check;--unwinding-assertions;--unwindset;memcpy_impl.0:3;--unwind;1;--object-bits;8"
```
Comment (markrtuttle) [inline .cbmc-batch/jobs/aws_array_list_back/cbmc-batch.yaml:2]: Object bits 8 bits is the default.

File: .cbmc-batch/jobs/aws_array_list_capacity/aws_array_list_capacity_harness.c [hunk 7]
```diff
@@ -16,36 +16,28 @@
...
 
-    struct aws_allocator *alloc = list->alloc;
-    size_t current_size = list->current_size;
-    size_t length = list->length;
-    size_t item_size = list->item_size;
-    void *data = list->data;
+    /* save current state of the data structure */
+    struct aws_array_list old = list;
+    struct store_byte_from_buffer old_byte;
+    save_byte_from_array((uint8_t *)list.data, list.current_size, &old_byte);
```
Comment (markrtuttle) [inline .cbmc-batch/jobs/aws_array_list_capacity/aws_array_list_capacity_harness.c:34]: This is a trick we use to prove no byte in the array has changed.  I don't think this will be clear to someone outside the proof community.  This is one thing that might be worth explaining in a top-level readme file somewhere.

File: .cbmc-batch/jobs/aws_array_list_clean_up/aws_array_list_clean_up_harness.c [hunk 8]
```diff
@@ -16,30 +16,26 @@
...
-    assert(list->current_size == 0);
-    assert(list->length == 0);
-    assert(list->item_size == 0);
-    assert(list->data == NULL);
+    assert(aws_array_list_is_valid(&list));
+    assert(list.alloc == NULL);
+    assert(list.current_size == 0);
+    assert(list.length == 0);
+    assert(list.item_size == 0);
+    assert(list.data == NULL);
```
Comment (markrtuttle) [inline .cbmc-batch/jobs/aws_array_list_clean_up/aws_array_list_clean_up_harness.c:40]: I feel surprised to be reading predicates with fairly descriptive names, and then to find this "null" predicate spelled out in the code in a style different from the others.  Not a big deal.

File: .cbmc-batch/jobs/aws_array_list_clear/aws_array_list_clear_harness.c [hunk 9]
```diff
@@ -16,35 +16,29 @@
...
-    assert(list->item_size == item_size);
-    assert(list->data == data);
+    /* assertions */
+    assert(aws_array_list_is_valid(&list));
+    (list.data) ? assert(list.length == 0) : assert(list.length == old.length);
+    assert(list.alloc == old.alloc);
+    assert(list.current_size == old.current_size);
+    assert(list.item_size == old.item_size);
+    assert(list.data == old.data);
 }
```
Comment (markrtuttle) [inline .cbmc-batch/jobs/aws_array_list_clear/aws_array_list_clear_harness.c:44]: I don't know the semantics of "clear", but this is a surprising list of assertions.  In particular, if there is list data, then the list length 0, but if there is no data then the list lengths are unchanged.

File: .cbmc-batch/jobs/aws_array_list_copy/aws_array_list_copy_harness.c [hunk 10]
```diff
@@ -16,61 +16,30 @@
...
-    assert(from->alloc == from_alloc);
-    assert(from->current_size == from_current_size);
-    assert(from->item_size == from_item_size);
-    assert(from->data == from_data);
+
+    /* assertions */
+    assert(aws_array_list_is_valid(&from));
+    assert(aws_array_list_is_valid(&to));
+    assert(from.item_size == to.item_size);
 }
```
Comment (markrtuttle) [inline .cbmc-batch/jobs/aws_array_list_copy/aws_array_list_copy_harness.c:45]: No check of the data copied?

File: source/byte_buf.c [hunk 11]
```diff
@@ -414,9 +414,7 @@ int aws_byte_buf_append_with_lookup(
         return AWS_OP_ERR;
     }
 
-    AWS_POSTCONDITION(aws_byte_buf_is_valid(to));
-    AWS_POSTCONDITION(aws_byte_cursor_is_valid(from));
-    return AWS_OP_SUCCESS;
+    return AWS_RETURN_SUCCESS_WITH_POSTCONDITION(aws_byte_cursor_is_valid(from) && aws_byte_buf_is_valid(to));
 }
```
Comment (markrtuttle) [inline source/byte_buf.c:418]: Is this macro really easier to read that the original code?  It is done, maybe not worth changing, but I would reconsider this style in future modules.

File: tests/CMakeLists.txt [hunk 12]
```diff
@@ -260,8 +260,11 @@ add_test_case(rfc822_invalid_format)
 add_test_case(rfc822_invalid_tz)
 add_test_case(rfc822_invalid_auto_format)
 add_test_case(iso8601_utc_parsing)
+add_test_case(iso8601_basic_utc_parsing)
 add_test_case(iso8601_utc_parsing_auto_detect)
```
Comment (markrtuttle) [inline tests/CMakeLists.txt:264]: The existence of the source/date_time.c code is pretty mysterious to me in a proof package.  It would probably help if there were documentation in the file that you are writing tests.

**Review by markrtuttle — APPROVED** [review]

File: .cbmc-batch/jobs/aws_array_list_capacity/aws_array_list_capacity_harness.c (hunk 5 above)
Comment (feliperodri) [reply .cbmc-batch/jobs/aws_array_list_capacity/aws_array_list_capacity_harness.c:20]: @danielsn @tautschnig any thoughts?
